### PR TITLE
Adding plan-service folder for master merge

### PR DESCRIPTION
**Pull request by Priyanka-eGov**
## Summary by CodeRabbit

- **New Features**
	- Introduced the Plan Service, offering REST API functionality to manage micro plan configurations and plans.
	- Added Kafka producers and consumers for real-time event processing in plan configurations and plans.
	- Included comprehensive API endpoints for creating, updating, and searching plan configurations and beneficiaries.
- **Documentation**
	- Added detailed README for the Plan Service with API contracts, endpoints, and dependency information.
- **Enhancements**
	- Integrated new dependencies for improved functionality and support: Spring Boot starters, Flyway DB, PostgreSQL, Swagger, Lombok, Jackson, and JUnit.
	- Implemented configuration settings for MDMS, persister topic, and plan-related topics.
	- Established database query builders and row mappers to support complex data retrieval and mapping for plans and configurations.
- **Backend Services**
	- Added comprehensive services for plan configuration management and validation.
	- Implemented repositories for handling database operations related to plans and plan configurations.

### CODE REVIEW

**Review by shashwat-egov — APPROVED** [review]

**Review by shashwat-egov — APPROVED** [review]

File: health-services/plan-service/src/main/resources/db/migrate.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+flyway -url=$DB_URL -table=$SCHEMA_TABLE -user=$FLYWAY_USER -password=$FLYWAY_PASSWORD -locations=$FLYWAY_LOCATIONS -baselineOnMigrate=true -outOfOrder=true -ignoreMissingMigrations=true migrate
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/resources/db/migrate.sh:3]: all commands are tested and working fine.

File: health-services/plan-service/src/main/java/digit/config/MainConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    public MappingJackson2HttpMessageConverter jacksonConverter(ObjectMapper objectMapper) {
+        MappingJackson2HttpMessageConverter converter = new MappingJackson2HttpMessageConverter();
+        converter.setObjectMapper(objectMapper);
+        return converter;
+    }
+
+    @Bean
+    public SingleColumnRowMapper singleColumnRowMapper() {
+        return new SingleColumnRowMapper<>(String.class);
+    }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/config/MainConfiguration.java:44]: Both approaches are correct and will produce the same configured ObjectMapper

File: health-services/plan-service/src/main/resources/db/migration/main/V20240305113045__plan_configuration_create_ddl.sql [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  id                            character varying(64),
+  mapped_from                   character varying(256),
+  mapped_to                     character varying(256),
+  plan_configuration_id         character varying(64),
+  created_by                    character varying(64),
+  created_time                  bigint,
+  last_modified_by              character varying(64),
+  last_modified_time            bigint,
+  CONSTRAINT uk_plan_configuration_mapping_id PRIMARY KEY (id)
+);
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/resources/db/migration/main/V20240305113045__plan_configuration_create_ddl.sql:73]: Not required. Db Diagram provided in readme

File: health-services/plan-service/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+### Kafka Consumers
+NA
+
+### Kafka Producers
+- plan-config-create-topic
+- plan-config-update-topic
+
+- save-plan
+- update-plan
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/README.md:43]: already there.

File: health-services/plan-service/src/main/java/digit/repository/rowmapper/PlanRowMapper.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+            }
+
+            addActivities(rs, planEntry, activityMap, conditionMap);
+            addResources(rs, planEntry, resourceMap);
+            addTargets(rs, planEntry, targetMap);
+            planMap.put(planId, planEntry);
+        }
+
+        return new ArrayList<>(planMap.values());
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/repository/rowmapper/PlanRowMapper.java:71]: this is required.

File: health-services/plan-service/pom.xml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        <dependency>
+            <groupId>org.egov.services</groupId>
+            <artifactId>tracer</artifactId>
+            <version>2.9.0-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+          <groupId>org.egov</groupId>
+          <artifactId>mdms-client</artifactId>
+          <version>2.9.0-SNAPSHOT</version>
+          <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/pom.xml:75]: SNAPSHOT versions are required.

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+        // Validate dependencies
+        validateActivityDependencies(request);
+
+        // Validate Target's Metrics against MDMS
+        validateTargetMetrics(request, mdmsData);
+
+        // Validate Metric Detail's Unit against MDMS
+        validateMetricDetailUnit(request, mdmsData);
+    }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:80]: That is being done.

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+            metricListFromMDMS = JsonPath.read(mdmsData, jsonPathForMetric);
+        } catch (Exception e) {
+            throw new CustomException(JSONPATH_ERROR_CODE, JSONPATH_ERROR_MESSAGE);
+        }
+
+        for (Target target : plan.getTargets()) {
+            if (!metricListFromMDMS.contains(target.getMetric())) {
+                throw new CustomException(METRIC_NOT_FOUND_IN_MDMS_CODE, METRIC_NOT_FOUND_IN_MDMS_MESSAGE);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:367]: The error message is informative.

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        Object mdmsData = mdmsUtil.fetchMdmsData(request.getRequestInfo(), rootTenantId);
+
+        validateAssumptionKeyAgainstMDMS(request, mdmsData);
+        validateAssumptionValue(planConfiguration);
+        validateFilestoreId(planConfiguration);
+        validateTemplateIdentifierAgainstMDMS(request, mdmsData);
+        validateOperationsInputAgainstMDMS(request, mdmsData);
+        validateResourceMappingAgainstMDMS(request, mdmsData);
+        validateMappedToUniqueness(planConfiguration.getResourceMapping());
+    }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:51]: That is being done.

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+        // Ensure at least one active file for each required template identifier
+        for (Object requiredTemplate : requiredTemplateIdentifierFromMDMS) {
+            if (!activeRequiredTemplates.contains(requiredTemplate)) {
+                log.error("Required Template Identifier " + requiredTemplate + " does not have any active file.");
+                throw new CustomException(REQUIRED_TEMPLATE_IDENTIFIER_NOT_FOUND_CODE, REQUIRED_TEMPLATE_IDENTIFIER_NOT_FOUND_MESSAGE);
+            }
+        }
+
+    }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:175]: Error message is informative.

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+        }
+
+        for(Assumption assumption : planConfiguration.getAssumptions())
+        {
+            if(!assumptionListFromMDMS.contains(assumption.getKey()))
+            {
+                log.error("Assumption Key " + assumption.getKey() + " is not present in MDMS");
+                throw new CustomException(ASSUMPTION_KEY_NOT_FOUND_IN_MDMS_CODE, ASSUMPTION_KEY_NOT_FOUND_IN_MDMS_MESSAGE);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:102]: Append path at the end of error message.

File: health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * This method emits an event to the persister for it to save the plan in the database.
+     * @param planRequest
+     */
+    @Override
+    public void create(PlanRequest planRequest) {
+        producer.push(config.getPlanCreateTopic(), planRequest);
+    }
```
[the remark this answers is not in the document]
Comment (Priyanka-eGov) [reply health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java:50]: not required.

File: health-services/plan-service/pom.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/pom.xml:75]: version 2.9.0 is not present and its needed to use snapshot

File: health-services/plan-service/src/main/java/digit/service/PlanConfigurationService.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+     * @param request The search request containing the criteria.
+     * @return A list of plan configurations that match the search criteria.
+     */
+    public PlanConfigurationResponse search(PlanConfigurationSearchRequest request) {
+        validator.validateSearchRequest(request);
+        PlanConfigurationResponse response = PlanConfigurationResponse.builder().
+                responseInfo(responseInfoFactory.createResponseInfoFromRequestInfo(request.getRequestInfo(), true))
+                .planConfiguration(repository.search(request.getPlanConfigurationSearchCriteria()))
+                .totalCount(repository.count(request.getPlanConfigurationSearchCriteria()))
+                .build();
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanConfigurationService.java:67]: Fixed

File: health-services/plan-service/src/main/java/digit/service/PlanEnricher.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        // Generate id for resources
+        body.getPlan().getResources().forEach(resource -> UUIDEnrichmentUtil.enrichRandomUuid(resource, "id"));
+
+        // Generate id for targets
+        body.getPlan().getTargets().forEach(target -> UUIDEnrichmentUtil.enrichRandomUuid(target, "id"));
+
+        // Enrich audit details
+        body.getPlan().setAuditDetails(AuditDetailsEnrichmentUtil
+                .prepareAuditDetails(body.getPlan().getAuditDetails(), body.getRequestInfo(), Boolean.TRUE));
+
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanEnricher.java:54]: fixed.

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+            }
+        });
+    }
+
+    /**
+     * This method validates if the execution plan id provided in the request exists
+     * @param request
+     */
+    private void validateExecutionPlanExistence(PlanRequest request) {
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:137]: removed

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+     * @param request The PlanConfigurationRequest to be enriched.
+     * @throws CustomException if user information is missing in the request.
+     */
+    public void enrichCreate(PlanConfigurationRequest request) {
+        enrichPlanConfiguration(request.getPlanConfiguration());
+        if (request.getRequestInfo().getUserInfo() == null)
+            throw new CustomException(USERINFO_MISSING_CODE, USERINFO_MISSING_MESSAGE);
+
+        enrichAuditDetails(request, Boolean.TRUE);
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:40]: fixed

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:102]: added that into code

File: health-services/plan-service/src/main/java/digit/util/MdmsUtil.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+        } catch (Exception e) {
+            log.error(ERROR_WHILE_FETCHING_FROM_MDMS, e);
+        }
+
+        Object result = mdmsResponse.getMdmsRes();
+        if (result == null || ObjectUtils.isEmpty(result)) {
+            log.error(NO_MDMS_DATA_FOUND_FOR_GIVEN_TENANT_MESSAGE + " - " + tenantId);
+            throw new CustomException(NO_MDMS_DATA_FOUND_FOR_GIVEN_TENANT_CODE, NO_MDMS_DATA_FOUND_FOR_GIVEN_TENANT_MESSAGE);
+        }
+        return result;
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/util/MdmsUtil.java:55]: name change

File: health-services/plan-service/src/main/java/digit/web/models/Operation.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        @JsonCreator
+        public static OperatorEnum fromValue(String text) {
+            for (OperatorEnum b : OperatorEnum.values()) {
+                if (String.valueOf(b.value).equals(text)) {
+                    return b;
+                }
+            }
+            return null;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/web/models/Operation.java:90]: added exception.

File: health-services/plan-service/src/main/java/digit/web/models/Plan.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    private List<Resource> resources;
+
+    @JsonProperty("targets")
+    @Valid
+    private List<Target> targets;
+
+    @JsonProperty("auditDetails")
+    private AuditDetails auditDetails = null;
+
+}
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/web/models/Plan.java:64]: Not needed.

File: health-services/plan-service/src/main/java/digit/web/models/PlanConfigurationSearchCriteria.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class PlanConfigurationSearchCriteria {
+
+    @JsonProperty("tenantId")
+    @Size(min = 1, max = 100)
+    @NotNull
+    private String tenantId = null;
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/web/models/PlanConfigurationSearchCriteria.java:28]: changes to @Size(min = 2, max = 256)

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 21]
```diff
@@ -0,0 +1,424 @@
...
+                .filter(Assumption::getActive)
+                .map(Assumption::getKey)
+                .collect(Collectors.toSet());
+
+        List<Operation> operations = planConfiguration.getOperations();
+        for (Operation operation : operations) {
+            // Check if the operation is using an assumption key that is not in the set of active assumption keys
+            if (operation.getActive() && !activeAssumptionKeys.contains(operation.getAssumptionValue())) {
+                log.error("Assumption Value " + operation.getAssumptionValue() + " is not present in the list of active Assumption Keys");
+                throw new CustomException(ASSUMPTION_VALUE_NOT_FOUND_CODE, ASSUMPTION_VALUE_NOT_FOUND_MESSAGE + " - " + operation.getAssumptionValue());
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:71]: fixed

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+     * This validation method validates if the dependent activities are valid and if they form a cycle
+     * @param request
+     */
+    private void validateActivityDependencies(PlanRequest request) {
+        // Check if dependent activity codes are valid
+        validateDependentActivityCodes(request);
+
+        // Check if dependent activities form a cycle
+        checkForCycleInActivityDependencies(request);
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:89]: fix

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 23]
```diff
@@ -0,0 +1,383 @@
...
+     * This method validates if the plan configuration id provided in the request exists
+     * @param request
+     */
+    private void validatePlanConfigurationExistence(PlanRequest request) {
+        // If plan id provided is invalid, throw an exception
+        if(!request.getPlan().getPlanConfigurationId().isEmpty() && CollectionUtils.isEmpty(planConfigurationRepository.search(PlanConfigurationSearchCriteria.builder()
+                .id(request.getPlan().getPlanConfigurationId())
+                .tenantId(request.getPlan().getTenantId())
+                .build()))) {
+            throw new CustomException(INVALID_PLAN_CONFIG_ID_CODE, INVALID_PLAN_CONFIG_ID_MESSAGE);
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:179]: fixed

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 24]
```diff
@@ -0,0 +1,383 @@
...
+            metricListFromMDMS = JsonPath.read(mdmsData, jsonPathForMetric);
+        } catch (Exception e) {
+            throw new CustomException(JSONPATH_ERROR_CODE, JSONPATH_ERROR_MESSAGE);
+        }
+
+        for (Target target : plan.getTargets()) {
+            if (!metricListFromMDMS.contains(target.getMetric())) {
+                throw new CustomException(METRIC_NOT_FOUND_IN_MDMS_CODE, METRIC_NOT_FOUND_IN_MDMS_MESSAGE);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:358]: fixed

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 25]
```diff
@@ -0,0 +1,424 @@
...
+     * @throws CustomException If a duplicate 'mappedTo' value is found.
+     */
+    public static void validateMappedToUniqueness(List<ResourceMapping> resourceMappings) {
+        Set<String> uniqueMappedToSet = new HashSet<>();
+        for (ResourceMapping mapping : resourceMappings) {
+            String uniqueKey = mapping.getFilestoreId() + "-" + mapping.getMappedTo();
+            if (!uniqueMappedToSet.add(uniqueKey)) {
+                log.error("Duplicate MappedTo " + mapping.getMappedTo() + " for FilestoreId " + mapping.getFilestoreId());
+                throw new CustomException(DUPLICATE_MAPPED_TO_VALIDATION_ERROR_CODE,
+                        DUPLICATE_MAPPED_TO_VALIDATION_ERROR_MESSAGE + " - " + mapping.getMappedTo() + " for FilestoreId " + mapping.getFilestoreId());
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:252]: na

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 26]
```diff
@@ -0,0 +1,383 @@
...
+            throw new CustomException("PLAN_ACTIVITIES_NOT_ALLOWED", "Activities are not allowed if execution plan id is provided");
+        }
+
+        // Validate activity dates
+        if(!CollectionUtils.isEmpty(request.getPlan().getActivities())) {
+            request.getPlan().getActivities().forEach(activity -> {
+                if(activity.getPlannedEndDate() < activity.getPlannedStartDate())
+                    throw new CustomException("INVALID_ACTIVITY_DATES", "Planned end date cannot be before planned start date");
+            });
+        }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanValidator.java:166]: fixed

File: health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+     * This method emits an event to the persister for it to save the plan in the database.
+     * @param planRequest
+     */
+    @Override
+    public void create(PlanRequest planRequest) {
+		try {
+			producer.push(config.getPlanCreateTopic(), planRequest);
+		} catch (Exception e) {
+			log.info("Pushing message to topic " + config.getPlanCreateTopic() + " failed.", e);
+		}
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java:53]: FIxed

File: health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+     * This method emits an event to the persister for it to update the plan in the database.
+     * @param planRequest
+     */
+    @Override
+	public void update(PlanRequest planRequest) {
+		try {
+			producer.push(config.getPlanUpdateTopic(), planRequest);
+		} catch (Exception e) {
+			log.info("Pushing message to topic " + config.getPlanUpdateTopic() + " failed.", e);
+		}
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/repository/impl/PlanRepositoryImpl.java:88]: Fixed

File: health-services/plan-service/src/main/java/digit/service/PlanEnricher.java [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+            }
+        });
+
+        // Set empty value in dependencies list when it is empty or null
+        body.getPlan().getActivities().forEach(activity -> {
+            if(CollectionUtils.isEmpty(activity.getDependencies())) {
+                List<String> emptyStringList = new ArrayList<>();
+                emptyStringList.add("");
+                activity.setDependencies(emptyStringList);
+            }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/PlanEnricher.java:45]: NA

File: health-services/plan-service/src/main/java/digit/repository/impl/PlanConfigurationRepositoryImpl.java [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    /**
+     * Pushes an updated existing plan configuration to persister kafka topic.
+     * @param planConfigurationRequest The request containing the updated plan configuration details.
+     */
+    @Override
+    public void update(PlanConfigurationRequest planConfigurationRequest) {
+        producer.push(config.getPlanConfigUpdateTopic(), planConfigurationRequest);
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/repository/impl/PlanConfigurationRepositoryImpl.java:98]: Fixed

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+
+        // Generate id for assumptions
+        planConfiguration.getAssumptions().forEach(assumption -> UUIDEnrichmentUtil.enrichRandomUuid(assumption, "id"));
+
+        // Generate id for operations
+        planConfiguration.getOperations().forEach(operation -> UUIDEnrichmentUtil.enrichRandomUuid(operation, "id"));
+
+        // Generate id for resource mappings
+        planConfiguration.getResourceMapping().forEach(resourceMapping -> UUIDEnrichmentUtil.enrichRandomUuid(resourceMapping, "id"));
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:67]: NA

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 32]
```diff
@@ -0,0 +1,186 @@
...
+    public void enrichActiveForResourceMapping(File file, List<ResourceMapping> resourceMappings) {
+        if (!file.getActive()) {
+            // Set all corresponding resource mappings to inactive
+            for (ResourceMapping mapping : resourceMappings) {
+                if (mapping.getFilestoreId().equals(file.getFilestoreId())) {
+                    mapping.setActive(false);
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:147]: NA

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 33]
```diff
@@ -0,0 +1,186 @@
...
+            if (ObjectUtils.isEmpty(assumption.getId())) {
+                assumption.setActive(Boolean.TRUE);
+            }
+        });
+
+        planConfiguration.getResourceMapping().forEach(resourceMapping -> {
+            if (ObjectUtils.isEmpty(resourceMapping.getId())) {
+                resourceMapping.setActive(Boolean.TRUE);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:183]: NA

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 34]
```diff
@@ -0,0 +1,186 @@
...
+    /**
+     * Enriches the PlanConfigurationRequest for creating a new plan configuration.
+     * This method enriches the plan configuration with generated IDs, validates user information, and enriches audit details for create operation.
+     * @param request The PlanConfigurationRequest to be enriched.
+     * @throws CustomException if user information is missing in the request.
+     */
+    public void enrichCreate(PlanConfigurationRequest request) {
+        enrichPlanConfiguration(request.getPlanConfiguration());
+        if(ObjectUtils.isEmpty(request.getRequestInfo().getUserInfo()))
+            throw new CustomException(USERINFO_MISSING_CODE, USERINFO_MISSING_MESSAGE);
```
[the remark this answers is not in the document]
Comment (devdatta-egov) [reply health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:37]: FIxed

**Review by kavi-egov — APPROVED** [review]

**Pull request merged**